Knesset Plenum protocols datasource processing

yoadsn opened the pull request:
This PR contributes 2 parts and modifies 3 others:
- NEW - "download Knesset recordings" module
- MODIFY - Splitting long recordings in a more robust and flexible way (still using Silero-VAD)
- MODIFY - Transcribe recording splits using OpenAI compatible transcription server
- MODIFY - Transcription server exposes OpenAI compatible transcription API
- NEW - "Align transcriptions" module to leverage existing VTT transcripts of an audio recording to improve split base transcriptions

yairl commented on .gitignore:
```diff
@@ -3,3 +3,18 @@ __pycache__
+# Temporary Scrips
+tmp_*.py
+
+# Local Configiuration files
```
configuration

yoadsn replied:
@yairl 
Making sure I got this one - you prefer to use "configuration" as the folder name and not the "config" shorthand.
Correct?

yairl replied:
It's a typo - Configiuration.
Just asked to fix it :)

yoadsn replied:
🤦 done.

yairl commented on align_transcriptions.py:
```diff
@@ -0,0 +1,559 @@
+        is_fallback = True
+
+    if not is_fallback:
+        qury_embedding = encode_text_stream(
```
qury -> query (if that's what you meant)

yoadsn replied:
Indeed - tried to use shorter name here - but saving one char looks like a typo indeed.

yairl commented on server/server_models.py:
```diff
@@ -0,0 +1,173 @@
+class CreateTranscriptionRequest(BaseModel):
+    file: bytes = Field(
+        ...,
+        description="The audio file object (not file name) to transcribe, in one of these formats: flac, mp3, mp4, mpeg, mpga, m4a, ogg, wav, or webm.",
```
Might want to have this non-OpenAI compatible and instead 'any format supported by ffmpeg'.

Not asking for a change in this commit.

yairl commented on server/transcribe-server.py:
```diff
@@ -1,42 +1,115 @@
+                "timestamp_granularities[]"
+            )
+    except Exception as e:
+        return str(e), 400
```
Nicer if you can use a named constant (example: HTTPStatus.TOO_MANY_REQUESTS for 429) in relevant places.

yairl commented on server/transcribe-server.py:
```diff
@@ -1,42 +1,115 @@
+        in_flight_semaphore.release()
+
+
+@app.route("/execute", methods=["POST"])
```
I think this path should be removed.

No one should be using it if we use the new format, and using two different endpoint types means more code to maintain and more questions to answer.

Your new code is the way to go.

yairl commented on transcribe.py:
```diff
@@ -1,102 +1,296 @@
+        "segments": [
+            {
+                "id": segment["id"],
+                "seek": segment.get("seek") or 0.0,
```
Why do we have all those 'or' cases?
When does that happen?

yoadsn replied:
I am ashamed to say I have no CUDA device, this makes "faster-whisper" super-slow-whisper for me. Thus, I transcribe using whisper.cpp, which might not have all of those in the response.
To make sure we serialize all properly, I had those safety Or's added so no casting errors during serialization to Parquet will throw. (See my other comment about Parquet usage)

yairl replied:
No worries, just add a comment :)

yairl reviewed (commented):
Some minor comments inside.

Three extra notes here:

1. I started reading transcribe.py and figured out it's a bigger change; will go through it tomorrow.

2. Following our discussion about process_streaming.py, I did a few experiments of loading large audio files as 16k mono samples into memory.

An hour is ~200MB, 20 hours are ~4GB. I was easily able to load 60 hours on my machine, and could've loaded much more.

I think anyone who wants to run VAD on a 20-hour recording should reasonably have 4GB of RAM to spare, and avoid having the extra code to handle streaming in our code.

On the other hand, I think it is not reasonable to expect that someone with a Mac or similar (or anyone who just didn't have the right package) will have to use a side branch, or other hacks.

I'd like to offer an alternative approach: we drop the streaming code, and instead run ffmpeg just prior to loading.
The migrated file, now 16k mono, easily fits into all of our systems, and no extra requirements are there.

What do you think?

3. black auto-formatting: after seeing more of it I've come to seriously hate it.
It reminded me that opinionated is great - when it's MY opinions, not someone else's.

To quote one of my favorite hero-s: that's the worst kind of discrimination - the kind against ME!

Frankly though, after testing black a bit, it's pretty nice :)
My only request is to change the line length to 120; we generally have wide screens nowadays, and it's easier on the eye for me.

No need to change now - I will reformat the entire code-base (not just your changes) and commit a reformat AFTER your PR is done.

yoadsn commented:
Thanks for the feedback!

## transcribe changes

Let me give some background on that to maybe save you some questions.
It looks like a lot of changes since I moved some code to the top (I follow the code style that first defines functions in the file before using them in subsequent functions.)
Since for me transcription is an over-night run, I implemented a "state run" using Pandas that dumps once in a while to a parquet file.
At the end - I repackage it in a compatible JSON file with existing structure. This made the process more robust to crashes.
Personally, I would always use something like parquet or hdf5 files to capture all artifacts - but of course I would promote such change without you really wanting to go that path.
It's using the OpenAI compatible API, keeping the ability to use the "homegrown API" just in case we need that back comp initially.
I also encapsulated the transformation of the API response to the expected format the current implementation dumps so the file contents would be compatible with older transcription artifacts.
Finally, I introduced an optional "sleep" between transcription batches so the machine would not overheat and have fans running all night - again - this is optional and convenience.

## process_streaming
Sounds good - I will take process_streaming as a base since it contains other things I think  could be useful to keep. Remove the sliding window approach and instead create a preprocess step that generates the file ussing ffmpeg that will not require mono/down-sampling.
Once I have that we can re-review to see if we can get rid of more code.
End game here is to replace "process" with that final iteration of "process_streaming" - Let me know if this works.
I am starting on that as soon as I get more keyboard time.

## Formatting Using Black
Done (For files I have touched/created)

yoadsn commented:
@yairl Please note "process_streaming" changes done.
- Once approved - this file will be deleted and the original "process" script will be replaced for a clean merge.
- Since I saw huge time waste when files in the proper 16K/mono format are being trans-coded by ffmpeg to the same format - I implemented ffprobe based detection of source format such that we can simply copy the file which takes almost no time at all. Hope you are ok with this extra code, I feel it can make a big difference in big batch processing.

yairl commented on process_streaming.py:
```diff
@@ -0,0 +1,271 @@
+            (split["start"] / SAMPLING_RATE, split["end"] / SAMPLING_RATE) for split in speech_timestamps
+        ]
+
+        store_splits(audio_file, canonical_splits, target_dir)
```
Splits should be based on the original file, not the transcoded one.

yairl commented on process_streaming.py:
```diff
@@ -0,0 +1,271 @@
+        pathlib.Path(target_dir).mkdir(parents=True, exist_ok=True)
+
+        audio_file_in_processing_format = target_dir / "full.mp3"
+        audio_info = prepare_audio_for_processing(audio_file, audio_file_in_processing_format)
```
Make sure to transcode even if full.mp3 already exists.

full-16k-mono.mp3?

yoadsn commented:
@yairl I believe all we covered is now done.
Let's give it another shot - and when we are good - can merge this in.
